include/rdma/fabric.h

```diff
@@ -465,6 +481,14 @@ struct fi_fabric_attr {
 	uint32_t		api_version;
 };
 
+struct fi_hmem_attr {
+	enum fi_hmem_iface		iface;
+	enum fi_hmem_attr_opt		api_permitted;
+	enum fi_hmem_attr_opt		use_p2p;
+	enum fi_hmem_attr_opt		use_dev_reg_copy;
+	struct fi_hmem_attr		*next;
+};
```

Comment:
I think these are more along the lines of a configuration than an attribute.  The app is wanting to configure the provider behavior here, versus discovering provider capabilities.

I want to avoid an embedded linked list.  It adds complexity, and it's unlikely that a system will realistically have more than 1 type of hmem installed anyway.

My first thought is this should somehow be linked to memory registration, since that's ultimately where the provider is making decisions on how to perform a data transfer.  And the above configuration settings are basically letting the provider know what sort of optimizations it can perform when a data buffer is located in hmem.

Maybe there should be a more involved set of APIs to query and configure memory registration related functionality.  I can see the MR cache being part of this.

Reply:
>I want to avoid an embedded linked list. It adds complexity, and it's unlikely that a system will realistically have more than 1 type of hmem installed anyway.

We were debating internally on that. If we are confident there wouldn't be > 1 hmem type, we can get rid of the linked list structure.

Reply:
We prefer to add it to fi_info so the user can see if these configurations are being used by running fi_info.

Reply:
These settings are configurations (restrictions) specified by the application, not the provider. They differ from the fi_info attributes in that regard.

Reply:
The application may not care about the implementation details and leave these configurations unspecified, then it is up to the provider to choose whether to use them and return in fi_info.

Reply:
The values make no sense for a provider to return: REQUIRED, PREFERRED, etc.  These are application restrictions on the provider implementation.  If the app doesn't specify a restriction, it doesn't care what the provider does.  Except that an admin might...  Some restrictions don't even make sense for some providers, except to disable the provider completely.

There are a much larger set of restrictions that an application may need to set.  HMEM settings, MR cache controls, use of CPU atomics, use of shared memory, eager message sizes, receive side buffering for unexpected messages...  Whether these restrictions are per provider, per endpoint, per domain, or global is unknown.

struct fi_info isn't the place for this.  Consider the proposal has a linked list of these restrictions.  This is input into fi_getinfo().  What is a provider supposed to do with this list?  Pick the one it likes?  Apply all of them?  How does it resolve conflicts?  What does a provider do if it uses shared memory for local communication, where a setting doesn't apply?

Conceptually, the application or an administrator is programming the provider implementation through some sort of configuration mechanism.  That's typically been done using environment variables, or setop() when done programmatically.  We don't want to link all these configuration values off of fi_info.

Reply:
>Conceptually, the application or an administrator is programming the provider implementation through some sort of configuration mechanism. That's typically been done using environment variables, or setop() when done programmatically. We don't want to link all these configuration values off of fi_info.

We always prefer a programmatical way to do such configuration instead of environment variables. I explained why `setopt` is not an ideal place either for such configuration  in the thread below.

>There are a much larger set of restrictions that an application may need to set. HMEM settings, MR cache controls, use of CPU atomics, use of shared memory, eager message sizes, receive side buffering for unexpected messages... Whether these restrictions are per provider, per endpoint, per domain, or global is unknown.

There are always adhoc or provider specific configuration that you don't want to involve in an API. What we are achieving is to address the common pain points in the FI_HMEM interface that all providers may share by introducing incremental changes to the interface.

Among the 3 attributes we are introducing, `use_p2p` and `api_permitted` are something AWS or HPE (@iziemba correct me if wrong) showed interest. `use_dev_reg_copy` is something on the data transfer details that we can totally cut.